File: pkg/mcp/adapter/adapter.go
package adapter

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"log"
	"os"
	"sync"

	"github.com/example/pipeline-engine/internal/engine"
	gosdk "github.com/example/pipeline-engine/pkg/sdk/go"
)

const (
	jsonRPCVersion = "2.0"

	errCodeParseError     = -32700
	errCodeInvalidRequest = -32600
	errCodeMethodNotFound = -32601
	errCodeInvalidParams  = -32602
	errCodeInternalError  = -32603
)

// EngineClient represents the subset of the Go SDK consumed by the MCP adapter.
type EngineClient interface {
	CreateJob(ctx context.Context, req engine.JobRequest) (*engine.Job, error)
	StreamJob(ctx context.Context, req engine.JobRequest) (<-chan engine.StreamingEvent, *engine.Job, error)
	GetJob(ctx context.Context, jobID string) (*engine.Job, error)
	CancelJob(ctx context.Context, jobID string, reason string) (*engine.Job, error)
	RerunJob(ctx context.Context, jobID string, payload gosdk.RerunRequest) (*engine.Job, error)
	UpsertProviderProfile(ctx context.Context, profile engine.ProviderProfile) error
	StreamExistingJob(ctx context.Context, jobID string, afterSeq uint64) (<-chan engine.StreamingEvent, error)
}

// SDKClient adapts the Go SDK client to EngineClient.
type SDKClient struct {
	client *gosdk.Client
}

// NewSDKClient wraps the Go SDK client for use with the adapter.
func NewSDKClient(client *gosdk.Client) *SDKClient {
	return &SDKClient{client: client}
}

func (c *SDKClient) CreateJob(ctx context.Context, req engine.JobRequest) (*engine.Job, error) {
	return c.client.CreateJob(ctx, req)
}

func (c *SDKClient) StreamJob(ctx context.Context, req engine.JobRequest) (<-chan engine.StreamingEvent, *engine.Job, error) {
	return c.client.StreamJobs(ctx, req)
}

func (c *SDKClient) GetJob(ctx context.Context, jobID string) (*engine.Job, error) {
	return c.client.GetJob(ctx, jobID)
}

func (c *SDKClient) CancelJob(ctx context.Context, jobID string, reason string) (*engine.Job, error) {
	return c.client.CancelJob(ctx, jobID, reason)
}

func (c *SDKClient) RerunJob(ctx context.Context, jobID string, payload gosdk.RerunRequest) (*engine.Job, error) {
	return c.client.RerunJob(ctx, jobID, payload)
}

func (c *SDKClient) UpsertProviderProfile(ctx context.Context, profile engine.ProviderProfile) error {
	return c.client.UpsertProviderProfile(ctx, profile)
}

func (c *SDKClient) StreamExistingJob(ctx context.Context, jobID string, afterSeq uint64) (<-chan engine.StreamingEvent, error) {
	if afterSeq > 0 {
		return c.client.StreamJobByID(ctx, jobID, afterSeq)
	}
	return c.client.StreamJobByID(ctx, jobID)
}

// Tool describes a MCP tool entry returned from tools/list.
type Tool struct {
	Name        string                 `json:"name"`
	Description string                 `json:"description"`
	InputSchema map[string]any         `json:"inputSchema,omitempty"`
	Metadata    map[string]interface{} `json:"metadata,omitempty"`
}

// Options configures the Adapter.
type Options struct {
	Client EngineClient
	Reader io.Reader
	Writer io.Writer
	Logger *log.Logger
	Tools  []Tool
}

// Adapter bridges MCP JSON-RPC traffic to the Pipeline Engine SDK.
type Adapter struct {
	client EngineClient
	reader io.Reader
	writer io.Writer
	logger *log.Logger
	tools  []Tool
	enc    *json.Encoder
	dec    *json.Decoder
	mu     sync.Mutex
}

// NewAdapter creates a new Adapter.
func NewAdapter(opts Options) *Adapter {
	reader := opts.Reader
	if reader == nil {
		reader = os.Stdin
	}
	writer := opts.Writer
	if writer == nil {
		writer = os.Stdout
	}
	logger := opts.Logger
	if logger == nil {
		logger = log.New(os.Stderr, "[mcp] ", log.LstdFlags)
	}
	tools := opts.Tools
	if len(tools) == 0 {
		tools = defaultTools()
	}
	return &Adapter{
		client: opts.Client,
		reader: reader,
		writer: writer,
		logger: logger,
		tools:  tools,
		enc:    json.NewEncoder(writer),
		dec:    json.NewDecoder(reader),
	}
}

// Run starts processing MCP JSON-RPC messages until the context is cancelled or EOF.
func (a *Adapter) Run(ctx context.Context) error {
	if a.client == nil {
		return errors.New("adapter client is nil")
	}
	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		default:
		}

		var req rpcRequest
		if err := a.dec.Decode(&req); err != nil {
			if errors.Is(err, io.EOF) {
				return nil
			}
			return err
		}
		if req.JSONRPC != "" && req.JSONRPC != jsonRPCVersion {
			a.respondError(req.ID, errCodeInvalidRequest, "invalid jsonrpc version", nil)
			continue
		}
		if req.Method == "" {
			a.respondError(req.ID, errCodeInvalidRequest, "method is required", nil)
			continue
		}

		switch req.Method {
		case "initialize":
			result := map[string]any{
				"protocolVersion": "0.1",
				"serverInfo": map[string]string{
					"name":    "pipeline-engine-mcp",
					"version": "0.1.0",
				},
				"capabilities": map[string]any{
					"prompts":      map[string]bool{"list": false},
					"resources":    map[string]bool{"list": false},
					"tools":        map[string]bool{"list": true, "call": true},
					"experimental": map[string]bool{},
				},
			}
			a.respondResult(req.ID, result)
		case "tools/list":
			result := map[string]any{"tools": a.tools}
			a.respondResult(req.ID, result)
		case "tools/call":
			a.handleToolCall(ctx, req)
		default:
			a.respondError(req.ID, errCodeMethodNotFound, "method not implemented", nil)
		}
	}
}

func (a *Adapter) handleToolCall(ctx context.Context, req rpcRequest) {
	if len(req.Params) == 0 {
		a.respondError(req.ID, errCodeInvalidParams, "missing params", nil)
		return
	}
	var params toolCallParams
	if err := json.Unmarshal(req.Params, &params); err != nil {
		a.respondError(req.ID, errCodeInvalidParams, "invalid params", err.Error())
		return
	}
	switch params.ToolName {
	case "startPipeline":
		a.handleStartPipeline(ctx, req.ID, params.Arguments)
	case "getJob":
		a.handleGetJob(ctx, req.ID, params.Arguments)
	case "cancelJob":
		a.handleCancelJob(ctx, req.ID, params.Arguments)
	case "rerunJob":
		a.handleRerunJob(ctx, req.ID, params.Arguments)
	case "upsertProviderProfile":
		a.handleUpsertProviderProfile(ctx, req.ID, params.Arguments)
	case "streamJob":
		a.handleStreamJob(ctx, req.ID, params.Arguments)
	default:
		a.respondError(req.ID, errCodeMethodNotFound, "tool not found", params.ToolName)
	}
}

func (a *Adapter) handleStartPipeline(ctx context.Context, id json.RawMessage, raw json.RawMessage) {
	var args startPipelineArgs
	if err := json.Unmarshal(raw, &args); err != nil {
		a.respondError(id, errCodeInvalidParams, "invalid arguments", err.Error())
		return
	}
	if args.PipelineType == "" {
		a.respondError(id, errCodeInvalidParams, "pipeline_type is required", nil)
		return
	}
	req := engine.JobRequest{
		PipelineType:  engine.PipelineType(args.PipelineType),
		Input:         args.Input,
		Mode:          args.Mode,
		FromStepID:    args.FromStepID,
		ReuseUpstream: args.ReuseUpstream,
	}
	if args.Stream {
		eventCh, job, err := a.client.StreamJob(ctx, req)
		if err != nil {
			a.respondError(id, errCodeInternalError, "stream job failed", err.Error())
			return
		}
		a.emitToolEvent("startPipeline", engine.StreamingEvent{
			Event: "job_queued",
			JobID: job.ID,
			Data:  job,
		})
		for evt := range eventCh {
			select {
			case <-ctx.Done():
				return
			default:
			}
			a.emitToolEvent("startPipeline", evt)
		}
		a.respondResult(id, map[string]any{
			"job": job,
		})
		return
	}
	job, err := a.client.CreateJob(ctx, req)
	if err != nil {
		a.respondError(id, errCodeInternalError, "create job failed", err.Error())
		return
	}
	a.respondResult(id, map[string]any{"job": job})
}

func (a *Adapter) handleGetJob(ctx context.Context, id json.RawMessage, raw json.RawMessage) {
	var args getJobArgs
	if err := json.Unmarshal(raw, &args); err != nil {
		a.respondError(id, errCodeInvalidParams, "invalid arguments", err.Error())
		return
	}
	if args.JobID == "" {
		a.respondError(id, errCodeInvalidParams, "job_id is required", nil)
		return
	}
	job, err := a.client.GetJob(ctx, args.JobID)
	if err != nil {
		a.respondError(id, errCodeInternalError, "get job failed", err.Error())
		return
	}
	a.respondResult(id, map[string]any{"job": job})
}

func (a *Adapter) handleCancelJob(ctx context.Context, id json.RawMessage, raw json.RawMessage) {
	var args cancelJobArgs
	if err := json.Unmarshal(raw, &args); err != nil {
		a.respondError(id, errCodeInvalidParams, "invalid arguments", err.Error())
		return
	}
	if args.JobID == "" {
		a.respondError(id, errCodeInvalidParams, "job_id is required", nil)
		return
	}
	job, err := a.client.CancelJob(ctx, args.JobID, args.Reason)
	if err != nil {
		a.respondError(id, errCodeInternalError, "cancel job failed", err.Error())
		return
	}
	a.respondResult(id, map[string]any{"job": job})
}

func (a *Adapter) handleRerunJob(ctx context.Context, id json.RawMessage, raw json.RawMessage) {
	var args rerunJobArgs
	if err := json.Unmarshal(raw, &args); err != nil {
		a.respondError(id, errCodeInvalidParams, "invalid arguments", err.Error())
		return
	}
	if args.JobID == "" {
		a.respondError(id, errCodeInvalidParams, "job_id is required", nil)
		return
	}
	req := gosdk.RerunRequest{
		FromStepID:    args.FromStepID,
		ReuseUpstream: args.ReuseUpstream,
		OverrideInput: args.OverrideInput,
	}
	job, err := a.client.RerunJob(ctx, args.JobID, req)
	if err != nil {
		a.respondError(id, errCodeInternalError, "rerun job failed", err.Error())
		return
	}
	a.respondResult(id, map[string]any{"job": job})
}

func (a *Adapter) handleUpsertProviderProfile(ctx context.Context, id json.RawMessage, raw json.RawMessage) {
	var profile engine.ProviderProfile
	if err := json.Unmarshal(raw, &profile); err != nil {
		a.respondError(id, errCodeInvalidParams, "invalid profile", err.Error())
		return
	}
	if profile.ID == "" {
		a.respondError(id, errCodeInvalidParams, "id is required", nil)
		return
	}
	if err := a.client.UpsertProviderProfile(ctx, profile); err != nil {
		a.respondError(id, errCodeInternalError, "upsert provider failed", err.Error())
		return
	}
	a.respondResult(id, map[string]any{"profile": profile})
}

func (a *Adapter) handleStreamJob(ctx context.Context, id json.RawMessage, raw json.RawMessage) {
	var args streamJobArgs
	if err := json.Unmarshal(raw, &args); err != nil {
		a.respondError(id, errCodeInvalidParams, "invalid arguments", err.Error())
		return
	}
	if args.JobID == "" {
		a.respondError(id, errCodeInvalidParams, "job_id is required", nil)
		return
	}
	eventCh, err := a.client.StreamExistingJob(ctx, args.JobID, args.AfterSeq)
	if err != nil {
		a.respondError(id, errCodeInternalError, "stream job failed", err.Error())
		return
	}
	for evt := range eventCh {
		select {
		case <-ctx.Done():
			return
		default:
		}
		a.emitToolEvent("streamJob", evt)
	}
	a.respondResult(id, map[string]any{"job_id": args.JobID})
}

func (a *Adapter) respondResult(id json.RawMessage, result interface{}) {
	if len(id) == 0 {
		return
	}
	resp := rpcResponse{
		JSONRPC: jsonRPCVersion,
		ID:      id,
		Result:  result,
	}
	a.mu.Lock()
	defer a.mu.Unlock()
	_ = a.enc.Encode(&resp)
}

func (a *Adapter) respondError(id json.RawMessage, code int, message string, data interface{}) {
	if len(id) == 0 {
		return
	}
	resp := rpcResponse{
		JSONRPC: jsonRPCVersion,
		ID:      id,
		Error: &rpcError{
			Code:    code,
			Message: message,
			Data:    data,
		},
	}
	a.mu.Lock()
	defer a.mu.Unlock()
	_ = a.enc.Encode(&resp)
}

func (a *Adapter) emitToolEvent(toolName string, evt engine.StreamingEvent) {
	notification := rpcNotification{
		JSONRPC: jsonRPCVersion,
		Method:  "tool_event",
		Params: map[string]any{
			"toolName": toolName,
			"event":    evt.Event,
			"kind":     classifyEventKind(evt.Event),
			"seq":      evt.Seq,
			"payload":  evt,
		},
	}
	a.mu.Lock()
	defer a.mu.Unlock()
	_ = a.enc.Encode(&notification)
}

func classifyEventKind(eventName string) string {
	switch eventName {
	case "provider_chunk":
		return "chunk"
	case "item_completed":
		return "result"
	case "error", "job_failed", "step_failed":
		return "error"
	default:
		return "status"
	}
}

type rpcRequest struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      json.RawMessage `json:"id"`
	Method  string          `json:"method"`
	Params  json.RawMessage `json:"params"`
}

type rpcResponse struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      json.RawMessage `json:"id"`
	Result  interface{}     `json:"result,omitempty"`
	Error   *rpcError       `json:"error,omitempty"`
}

type rpcNotification struct {
	JSONRPC string      `json:"jsonrpc"`
	Method  string      `json:"method"`
	Params  interface{} `json:"params,omitempty"`
}

type rpcError struct {
	Code    int         `json:"code"`
	Message string      `json:"message"`
	Data    interface{} `json:"data,omitempty"`
}

type toolCallParams struct {
	ToolName  string          `json:"toolName"`
	Arguments json.RawMessage `json:"arguments"`
}

type startPipelineArgs struct {
	PipelineType  string          `json:"pipeline_type"`
	Input         engine.JobInput `json:"input"`
	Mode          string          `json:"mode,omitempty"`
	Stream        bool            `json:"stream,omitempty"`
	FromStepID    *engine.StepID  `json:"from_step_id,omitempty"`
	ReuseUpstream bool            `json:"reuse_upstream,omitempty"`
}

type getJobArgs struct {
	JobID string `json:"job_id"`
}

type cancelJobArgs struct {
	JobID  string `json:"job_id"`
	Reason string `json:"reason,omitempty"`
}

type rerunJobArgs struct {
	JobID         string           `json:"job_id"`
	FromStepID    *engine.StepID   `json:"from_step_id,omitempty"`
	ReuseUpstream bool             `json:"reuse_upstream,omitempty"`
	OverrideInput *engine.JobInput `json:"override_input,omitempty"`
}

type streamJobArgs struct {
	JobID    string `json:"job_id"`
	AfterSeq uint64 `json:"after_seq,omitempty"`
}

func defaultTools() []Tool {
	return []Tool{
		{
			Name:        "startPipeline",
			Description: "Create a new pipeline job via /v1/jobs",
			InputSchema: map[string]any{
				"type": "object",
				"required": []string{
					"pipeline_type",
					"input",
				},
				"properties": map[string]any{
					"pipeline_type": map[string]string{"type": "string"},
					"input":         map[string]string{"type": "object"},
					"mode":          map[string]string{"type": "string"},
					"stream":        map[string]string{"type": "boolean"},
				},
			},
		},
		{
			Name:        "getJob",
			Description: "Fetch job details via /v1/jobs/{id}",
			InputSchema: map[string]any{
				"type":       "object",
				"required":   []string{"job_id"},
				"properties": map[string]any{"job_id": map[string]string{"type": "string"}},
			},
		},
		{
			Name:        "cancelJob",
			Description: "Cancel a running job via /v1/jobs/{id}/cancel",
			InputSchema: map[string]any{
				"type":       "object",
				"required":   []string{"job_id"},
				"properties": map[string]any{"job_id": map[string]string{"type": "string"}, "reason": map[string]string{"type": "string"}},
			},
		},
		{
			Name:        "rerunJob",
			Description: "Rerun a job via /v1/jobs/{id}/rerun",
			InputSchema: map[string]any{
				"type":     "object",
				"required": []string{"job_id"},
				"properties": map[string]any{
					"job_id":         map[string]string{"type": "string"},
					"from_step_id":   map[string]string{"type": "string"},
					"reuse_upstream": map[string]string{"type": "boolean"},
				},
			},
		},
		{
			Name:        "upsertProviderProfile",
			Description: "Update ProviderProfile via /v1/config/providers",
			InputSchema: map[string]any{
				"type":     "object",
				"required": []string{"id", "kind"},
				"properties": map[string]any{
					"id":            map[string]string{"type": "string"},
					"kind":          map[string]string{"type": "string"},
					"base_uri":      map[string]string{"type": "string"},
					"api_key":       map[string]string{"type": "string"},
					"default_model": map[string]string{"type": "string"},
				},
			},
		},
		{
			Name:        "streamJob",
			Description: "Stream NDJSON events for an existing job via /v1/jobs/{id}/stream",
			InputSchema: map[string]any{
				"type":       "object",
				"required":   []string{"job_id"},
				"properties": map[string]any{"job_id": map[string]string{"type": "string"}},
			},
		},
	}
}

File: pkg/sdk/typescript/src/mcp/adapter.ts
import { createInterface } from "node:readline";
import type { Readable, Writable } from "node:stream";
import type { PipelineEngineClient } from "../client.js";
import type {
  Job,
  JobRequest,
  ProviderProfileInput,
  StreamingEvent
} from "../types.js";

const JSONRPC_VERSION = "2.0";

interface RPCRequest {
  jsonrpc?: string;
  id?: number | string | null;
  method: string;
  params?: any;
}

interface ToolCallParams {
  toolName: string;
  arguments?: any;
}

export interface MCPClient {
  createJob(req: JobRequest): Promise<Job>;
  streamJobs(req: JobRequest): Promise<{ job: Job; events: AsyncIterable<StreamingEvent> }>;
  getJob(jobID: string): Promise<Job>;
  cancelJob(jobID: string, reason?: string): Promise<Job>;
  rerunJob(jobID: string, payload: Record<string, unknown>): Promise<Job>;
  upsertProviderProfile(profile: ProviderProfileInput): Promise<void>;
  streamJobByID(jobID: string, afterSeq?: number): Promise<AsyncIterable<StreamingEvent>>;
}

export interface ToolDefinition {
  name: string;
  description: string;
  inputSchema?: Record<string, unknown>;
  metadata?: Record<string, unknown>;
}

export interface MCPAdapterOptions {
  client: MCPClient;
  input?: Readable;
  output?: Writable;
  tools?: ToolDefinition[];
  logger?: { error: (...args: unknown[]) => void };
}

export class MCPAdapter {
  private readonly client: MCPClient;
  private readonly input: Readable;
  private readonly output: Writable;
  private readonly tools: ToolDefinition[];
  private readonly logger?: { error: (...args: unknown[]) => void };

  constructor(options: MCPAdapterOptions) {
    if (!options?.client) {
      throw new Error("client is required");
    }
    this.client = options.client;
    this.input = options.input ?? (process.stdin as Readable);
    this.output = options.output ?? (process.stdout as Writable);
    this.tools = options.tools ?? defaultTools();
    this.logger = options.logger;
  }

  async run(): Promise<void> {
    const rl = createInterface({ input: this.input, crlfDelay: Infinity });
    try {
      for await (const rawLine of rl) {
        const line = rawLine.trim();
        if (!line) {
          continue;
        }
        let req: RPCRequest;
        try {
          req = JSON.parse(line) as RPCRequest;
        } catch (err) {
          this.respondError(null, -32700, "parse error", String(err));
          continue;
        }
        await this.handleRequest(req).catch((err) => {
          this.logger?.error?.("adapter error", err);
          this.respondError(req.id ?? null, -32603, "internal error", err instanceof Error ? err.message : String(err));
        });
      }
    } finally {
      rl.close();
    }
  }

  private async handleRequest(req: RPCRequest): Promise<void> {
    if (req.jsonrpc && req.jsonrpc !== JSONRPC_VERSION) {
      this.respondError(req.id ?? null, -32600, "invalid jsonrpc version", null);
      return;
    }
    if (!req.method) {
      this.respondError(req.id ?? null, -32600, "method is required", null);
      return;
    }
    switch (req.method) {
      case "initialize":
        this.respondResult(req.id, {
          protocolVersion: "0.1",
          serverInfo: { name: "pipeline-engine-mcp-ts", version: "0.1.0" },
          capabilities: {
            tools: { list: true, call: true },
            prompts: { list: false },
            resources: { list: false },
            experimental: {}
          }
        });
        break;
      case "tools/list":
        this.respondResult(req.id, { tools: this.tools });
        break;
      case "tools/call":
        await this.handleToolCall(req);
        break;
      default:
        this.respondError(req.id ?? null, -32601, "method not found", req.method);
    }
  }

  private async handleToolCall(req: RPCRequest): Promise<void> {
    if (!req.params) {
      this.respondError(req.id ?? null, -32602, "missing params", null);
      return;
    }
    const params = req.params as ToolCallParams;
    if (!params.toolName) {
      this.respondError(req.id ?? null, -32602, "toolName is required", null);
      return;
    }
    switch (params.toolName) {
      case "startPipeline":
        await this.handleStartPipeline(req.id, params.arguments ?? {});
        break;
      case "getJob":
        await this.handleGetJob(req.id, params.arguments ?? {});
        break;
      case "cancelJob":
        await this.handleCancelJob(req.id, params.arguments ?? {});
        break;
      case "rerunJob":
        await this.handleRerunJob(req.id, params.arguments ?? {});
        break;
      case "upsertProviderProfile":
        await this.handleUpsertProviderProfile(req.id, params.arguments ?? {});
        break;
      case "streamJob":
        await this.handleStreamJob(req.id, params.arguments ?? {});
        break;
      default:
        this.respondError(req.id ?? null, -32601, "tool not found", params.toolName);
    }
  }

  private async handleStartPipeline(id: RPCRequest["id"], args: any): Promise<void> {
    if (!args?.pipeline_type || !args?.input) {
      this.respondError(id ?? null, -32602, "pipeline_type and input are required", null);
      return;
    }
    const req: JobRequest = {
      pipeline_type: args.pipeline_type,
      input: args.input,
      mode: args.mode,
      from_step_id: args.from_step_id,
      reuse_upstream: args.reuse_upstream
    };
    if (args.stream) {
      const { job, events } = await this.client.streamJobs(req);
      this.emitToolEvent("startPipeline", { event: "job_queued", job_id: job.id, data: job });
      for await (const evt of events) {
        this.emitToolEvent("startPipeline", evt);
      }
      this.respondResult(id, { job });
      return;
    }
    const job = await this.client.createJob(req);
    this.respondResult(id, { job });
  }

  private async handleGetJob(id: RPCRequest["id"], args: any): Promise<void> {
    if (!args?.job_id) {
      this.respondError(id ?? null, -32602, "job_id is required", null);
      return;
    }
    const job = await this.client.getJob(args.job_id);
    this.respondResult(id, { job });
  }

  private async handleCancelJob(id: RPCRequest["id"], args: any): Promise<void> {
    if (!args?.job_id) {
      this.respondError(id ?? null, -32602, "job_id is required", null);
      return;
    }
    const job = await this.client.cancelJob(args.job_id, args.reason);
    this.respondResult(id, { job });
  }

  private async handleRerunJob(id: RPCRequest["id"], args: any): Promise<void> {
    if (!args?.job_id) {
      this.respondError(id ?? null, -32602, "job_id is required", null);
      return;
    }
    const payload: Record<string, unknown> = {};
    if (args.from_step_id) {
      payload.from_step_id = args.from_step_id;
    }
    if (typeof args.reuse_upstream === "boolean") {
      payload.reuse_upstream = args.reuse_upstream;
    }
    if (args.override_input) {
      payload.override_input = args.override_input;
    }
    const job = await this.client.rerunJob(args.job_id, payload);
    this.respondResult(id, { job });
  }

  private async handleUpsertProviderProfile(id: RPCRequest["id"], args: any): Promise<void> {
    if (!args?.id || !args?.kind) {
      this.respondError(id ?? null, -32602, "id and kind are required", null);
      return;
    }
    await this.client.upsertProviderProfile(args as ProviderProfileInput);
    this.respondResult(id, { profile: args });
  }

  private async handleStreamJob(id: RPCRequest["id"], args: any): Promise<void> {
    if (!args?.job_id) {
      this.respondError(id ?? null, -32602, "job_id is required", null);
      return;
    }
    const afterSeq =
      typeof args.after_seq === "number" && args.after_seq > 0 ? args.after_seq : undefined;
    const events = await this.client.streamJobByID(args.job_id, afterSeq);
    for await (const evt of events) {
      this.emitToolEvent("streamJob", evt);
    }
    this.respondResult(id, { job_id: args.job_id });
  }

  private respondResult(id: RPCRequest["id"], result: any): void {
    if (id === undefined || id === null) {
      return;
    }
    this.writeJSON({ jsonrpc: JSONRPC_VERSION, id, result });
  }

  private respondError(id: RPCRequest["id"], code: number, message: string, data: unknown): void {
    if (id === undefined || id === null) {
      return;
    }
    this.writeJSON({ jsonrpc: JSONRPC_VERSION, id, error: { code, message, data } });
  }

  private emitToolEvent(toolName: string, evt: StreamingEvent): void {
    const notification = {
      jsonrpc: JSONRPC_VERSION,
      method: "tool_event",
      params: {
        toolName,
        event: evt.event,
        kind: classifyEventKind(evt.event),
        seq: evt.seq,
        payload: evt
      }
    };
    this.writeJSON(notification);
  }

  private writeJSON(payload: unknown): void {
    try {
      this.output.write(`${JSON.stringify(payload)}\n`);
    } catch (err) {
      this.logger?.error?.("failed to write json", err);
    }
  }
}

function classifyEventKind(eventName: string): string {
  switch (eventName) {
    case "provider_chunk":
      return "chunk";
    case "item_completed":
      return "result";
    case "error":
    case "job_failed":
    case "step_failed":
      return "error";
    default:
      return "status";
  }
}

function defaultTools(): ToolDefinition[] {
  return [
    {
      name: "startPipeline",
      description: "Create a job via /v1/jobs",
      inputSchema: {
        type: "object",
        required: ["pipeline_type", "input"],
        properties: {
          pipeline_type: { type: "string" },
          input: { type: "object" },
          mode: { type: "string" },
          stream: { type: "boolean" }
        }
      }
    },
    {
      name: "streamJob",
      description: "Stream events for an existing job",
      inputSchema: {
        type: "object",
        required: ["job_id"],
        properties: {
          job_id: { type: "string" },
          after_seq: { type: "number", minimum: 0 }
        }
      }
    },
    {
      name: "getJob",
      description: "Fetch job details",
      inputSchema: {
        type: "object",
        required: ["job_id"],
        properties: { job_id: { type: "string" } }
      }
    },
    {
      name: "cancelJob",
      description: "Cancel a running job",
      inputSchema: {
        type: "object",
        required: ["job_id"],
        properties: {
          job_id: { type: "string" },
          reason: { type: "string" }
        }
      }
    },
    {
      name: "rerunJob",
      description: "Rerun a job from a specific step",
      inputSchema: {
        type: "object",
        required: ["job_id"],
        properties: {
          job_id: { type: "string" },
          from_step_id: { type: "string" },
          reuse_upstream: { type: "boolean" },
          override_input: { type: "object" }
        }
      }
    },
    {
      name: "upsertProviderProfile",
      description: "Upsert a provider profile",
      inputSchema: {
        type: "object",
        required: ["id", "kind"],
        properties: {
          id: { type: "string" },
          kind: { type: "string" },
          base_uri: { type: "string" },
          api_key: { type: "string" },
          default_model: { type: "string" }
        }
      }
    }
  ];
}

export function createMCPAdapterClient(client: PipelineEngineClient): MCPClient {
  return {
    createJob: (req) => client.createJob(req),
    streamJobs: (req) => client.streamJobs(req),
    getJob: (id) => client.getJob(id),
    cancelJob: (id, reason) => client.cancelJob(id, reason),
    rerunJob: (id, payload) => client.rerunJob(id, payload),
    upsertProviderProfile: (profile) => client.upsertProviderProfile(profile),
    streamJobByID: (id, afterSeq) => client.streamJobByID(id, afterSeq)
  };
}
